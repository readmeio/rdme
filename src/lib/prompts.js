const semver = require('semver');
const { prompt } = require('enquirer');
const parse = require('parse-link-header');

exports.generatePrompts = (versionList, selectOnly = false) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to use an existing project version or create a new one?',
    skip() {
      return selectOnly;
    },
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return selectOnly ? false : this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return selectOnly ? true : this.enquirer.answers.option === 'update';
    },
    hint: '1.0.0',
  },
];

function specOptions(specList, parsedDocs, currPage, totalPages) {
  const specs = specList.map(s => {
    return {
      message: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs.prev.page) specs.push({ message: `< Prev (page ${currPage - 1} of ${totalPages})`, value: 'prev' });
  if (parsedDocs.next.page) {
    specs.push({ message: `Next (page ${currPage + 1} of ${totalPages}) >`, value: 'next' });
  }
  return specs;
}

const updateOasPrompt = (specList, parsedDocs, currPage, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async result(spec) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      return spec;
    },
  },
];

exports.createOasPrompt = (specList, parsedDocs, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to update an existing OAS file or create a new one?',
    choices: [
      { message: 'Update existing', value: 'update' },
      { message: 'Create a new spec', value: 'create' },
    ],
    async result(picked) {
      if (picked === 'update') {
        try {
          const { specId } = await prompt(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        } catch (e) {
          return null;
        }
      }
      return picked;
    },
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    initial: opts.newVersion || false,
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
    validate(val) {
      return semver.valid(semver.coerce(val)) ? true : this.styles.danger('Please specify a semantic version.');
    },
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || isUpdate?.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
