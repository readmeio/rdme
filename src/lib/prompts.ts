import type { Version } from '../cmds/versions';
import type { Response } from 'node-fetch';
import type { Choice, PromptObject } from 'prompts';

import parse from 'parse-link-header';
import semver from 'semver';

import promptTerminal from './promptWrapper';
import { handleRes } from './readmeAPIFetch';

interface Spec {
  _id: string;
  title: string;
}

type SpecList = Spec[];

interface ParsedDocs {
  next?: {
    page: number;
    url: string;
  };
  prev?: {
    page: number;
    url: string;
  };
}

function specOptions(specList: SpecList, parsedDocs: ParsedDocs, currPage: number, totalPages: number): Choice[] {
  const specs = specList.map(s => {
    return {
      description: `API Definition ID: ${s._id}`, // eslint-disable-line no-underscore-dangle
      title: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs?.prev?.page) {
    specs.push({
      description: 'Go to the previous page',
      title: `< Prev (page ${currPage - 1} of ${totalPages})`,
      value: 'prev',
    });
  }
  if (parsedDocs?.next?.page) {
    specs.push({
      description: 'Go to the next page',
      title: `Next (page ${currPage + 1} of ${totalPages}) >`,
      value: 'next',
    });
  }
  return specs;
}

const updateOasPrompt = (
  specList: SpecList,
  parsedDocs: ParsedDocs,
  currPage: number,
  totalPages: number,
  getSpecs: (url: string) => Promise<Response>,
): PromptObject[] => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async format(spec: string) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      } else if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      }

      return spec;
    },
  },
];

export function createOasPrompt(
  specList: SpecList,
  parsedDocs: ParsedDocs,
  totalPages: number,
  getSpecs: ((url: string) => Promise<Response>) | null,
): PromptObject[] {
  return [
    {
      type: 'select',
      name: 'option',
      message: 'Would you like to update an existing OAS file or create a new one?',
      choices: [
        { title: 'Update existing', value: 'update' },
        { title: 'Create a new spec', value: 'create' },
      ],
      async format(picked: 'update' | 'create') {
        if (picked === 'update') {
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        }

        return picked;
      },
    },
  ];
}

/**
 * Series of prompts to construct a version object,
 * used in our `versions:create` and `versions:update` commands
 */
export function versionPrompt(
  /** list of versions, used for prompt about which version to fork */
  versionList: Version[],
  /** existing version if we're performing an update */
  isUpdate?: {
    is_stable: boolean;
  },
): PromptObject[] {
  return [
    {
      // only runs for versions:create command
      type: isUpdate ? null : 'select',
      name: 'from',
      message: 'Which version would you like to fork from?',
      choices: versionList.map(v => {
        return {
          title: v.version,
          value: v.version,
        };
      }),
    },
    {
      // only runs for versions:update command
      type: !isUpdate ? null : 'text',
      name: 'newVersion',
      message: 'What should the version be renamed to?',
      hint: '1.0.0',
      validate(val: string) {
        // allow empty string, in which case the version won't be renamed
        if (!val) return true;
        return semver.valid(semver.coerce(val)) ? true : 'Please specify a semantic version.';
      },
    },
    {
      // if the existing version being updated is already the main version,
      // we can't switch that so we skip this question
      type: isUpdate?.is_stable ? null : 'confirm',
      name: 'is_stable',
      message: 'Would you like to make this version the main version for this project?',
    },
    {
      type: 'confirm',
      name: 'is_beta',
      message: 'Should this version be in beta?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be hidden.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_public',
      message: 'Would you like to make this version public?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be deprecated.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_deprecated',
      message: 'Would you like to deprecate this version?',
    },
  ];
}
