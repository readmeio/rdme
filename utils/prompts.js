exports.generatePrompts = versionList => [
  {
    type: 'select',
    name: 'option',
    message:
      "We couldn't find a version in ReadMe matching the version in your OAS file. Would you like to use an existing version or create a new one?",
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || (isUpdate && isUpdate.is_stable);
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
