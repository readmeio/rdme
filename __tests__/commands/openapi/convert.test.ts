import type { OASDocument } from 'oas/types';

import fs from 'node:fs';

import prompts from 'prompts';
import { describe, it, expect, vi, beforeAll, beforeEach, afterEach, type MockInstance } from 'vitest';

import Command from '../../../src/commands/openapi/convert.js';
import { runCommandAndReturnResult } from '../../helpers/oclif.js';

const successfulConversion = () => 'Your API definition has been converted and bundled and saved to output.json!';

describe('rdme openapi convert', () => {
  let fsWriteFileSyncSpy: MockInstance<typeof fs.writeFileSync>;
  let reducedSpec: OASDocument;
  let run: (args?: string[]) => Promise<string>;
  let testWorkingDir: string;

  beforeAll(() => {
    run = runCommandAndReturnResult(Command);
  });

  beforeEach(() => {
    testWorkingDir = process.cwd();

    fsWriteFileSyncSpy = vi.spyOn(fs, 'writeFileSync').mockImplementationOnce((filename, data) => {
      reducedSpec = JSON.parse(data as string);
    });
  });

  afterEach(() => {
    process.chdir(testWorkingDir);
    vi.restoreAllMocks();
  });

  describe('converting', () => {
    it.each([
      ['Swagger 2.0', 'json', '2.0'],
      ['Swagger 2.0', 'yaml', '2.0'],
    ])('should support reducing a %s definition (format: %s)', async (_, format, specVersion) => {
      const spec = require.resolve(`@readme/oas-examples/${specVersion}/${format}/petstore-simple.${format}`);

      prompts.inject(['output.json']);

      await expect(run([spec])).resolves.toBe(successfulConversion());

      expect(fsWriteFileSyncSpy).toHaveBeenCalledWith('output.json', expect.any(String));
      expect(reducedSpec.tags).toHaveLength(1);
      expect(Object.keys(reducedSpec.paths)).toStrictEqual(['/pet/{petId}']);
      expect(Object.keys(reducedSpec.paths['/pet/{petId}'])).toStrictEqual(['get', 'post', 'delete']);
    });
  });

  it('should convert with no prompts via opts', async () => {
    const spec = 'petstore-simple.json';

    await expect(
      run([
        spec,
        '--workingDirectory',
        require.resolve(`@readme/oas-examples/2.0/json/${spec}`).replace(spec, ''),
        '--out',
        'output.json',
      ]),
    ).resolves.toBe(successfulConversion());

    expect(fsWriteFileSyncSpy).toHaveBeenCalledWith('output.json', expect.any(String));
    expect(Object.keys(reducedSpec.paths)).toStrictEqual(['/pet/{petId}']);
    expect(Object.keys(reducedSpec.paths['/pet/{petId}'])).toStrictEqual(['get', 'post', 'delete']);
  });

  describe('error handling', () => {
    it.each([['json'], ['yaml']])('should fail if given an OpenAPI 3.0 definition (format: %s)', async format => {
      const spec = require.resolve(`@readme/oas-examples/3.0/${format}/petstore.${format}`);

      const consoleWarnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});

      prompts.inject(['output.json']);

      await expect(run([spec])).resolves.toBe(successfulConversion());

      expect(fsWriteFileSyncSpy).toHaveBeenCalledWith('output.json', expect.any(String));
      expect(reducedSpec.tags).toHaveLength(3);
      expect(Object.keys(reducedSpec.paths)).toStrictEqual([
        '/pet',
        '/pet/findByStatus',
        '/pet/findByTags',
        '/pet/{petId}',
        '/pet/{petId}/uploadImage',
        '/store/inventory',
        '/store/order',
        '/store/order/{orderId}',
        '/user',
        '/user/createWithArray',
        '/user/createWithList',
        '/user/login',
        '/user/logout',
        '/user/{username}',
      ]);
      expect(Object.keys(reducedSpec.paths['/pet/{petId}'])).toStrictEqual(['get', 'post', 'delete']);

      expect(consoleWarnSpy).toHaveBeenCalledWith(
        '⚠️  Warning! The input file is already OpenAPI, so no conversion is necessary. Any external references will be bundled.',
      );

      consoleWarnSpy.mockRestore();
    });
  });
});
