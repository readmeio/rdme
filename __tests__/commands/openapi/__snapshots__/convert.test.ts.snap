// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rdme openapi convert > error handling > should warn if given an OpenAPI 3.0 definition (format: json) 1`] = `
{
  "result": "Your API definition has been converted and bundled and saved to output.json!",
  "stderr": "- Validating the API definition located at petstore.json...
⚠️  Warning! The input file is already OpenAPI, so no conversion is necessary. Any external references will be bundled.
",
  "stdout": "",
}
`;

exports[`rdme openapi convert > error handling > should warn if given an OpenAPI 3.0 definition (format: yaml) 1`] = `
{
  "result": "Your API definition has been converted and bundled and saved to output.json!",
  "stderr": "- Validating the API definition located at petstore.yaml...
⚠️  Warning! The input file is already OpenAPI, so no conversion is necessary. Any external references will be bundled.
",
  "stdout": "",
}
`;

exports[`rdme openapi convert > should convert with no prompts via opts 1`] = `
{
  "result": "Your API definition has been converted and bundled and saved to output.json!",
  "stderr": "- Validating the API definition located at petstore-simple.json...
",
  "stdout": "",
}
`;
